WT-12657 Add checkpoint cleanup utility thread (7.0 backport)

kommiharibabu opened the pull request:
Move the checkpoint cleanup operation as part of the checkpoint process into a utility thread to improve the checkpoint performance. Improving the checkpoint performance can improve the overall performance of the database server.

(cherry picked from commit 7d0407c2904a4d1638478e459faa1157da4f8ef2)

kommiharibabu committed:
WT-12657 Add checkpoint cleanup utility thread (#10366)

Move the checkpoint cleanup operation as part of the checkpoint process
into an utility thread to improve the checkpoint performance. Improving
the checkpoint performance can improve the overall performance of the
database server.

(cherry picked from commit 7d0407c2904a4d1638478e459faa1157da4f8ef2)

Commit:
fix the test

Commit:
s_all

Commit:
fix the test stat

kommiharibabu commented:
@etienneptl This PR will not be merged into 7.0 branch. I will run a couple of patch builds to confirm it is of good quality.

etienneptl commented:
Reviewed the PR and communicated feedback. Looks good! Thanks for putting it together.

Commit:
Merge branch 'mongodb-7.0' into wt-12657-v7.0-backport

Commit:
Bring the WT-12883 change to avoid logging error messages of dhandle not found

kommiharibabu committed:
Merge branch 'mongodb-7.0' into wt-12657-v7.0-backport

etienneptl committed:
Merge branch 'mongodb-7.0' into wt-12657-v7.0-backport

kommiharibabu committed:
Merge branch 'mongodb-7.0' into wt-12657-v7.0-backport-2

kommiharibabu committed:
fix the new stat changes

kommiharibabu committed:
WT-12657 Add use_thread config to make checkpoint cleanup thread opt-in

Add a checkpoint_cleanup.use_thread boolean config (default false) so
the dedicated cleanup utility thread can be enabled selectively. When
false, obsolete page cleanup runs inline during the checkpoint tree walk
(original pre-WT-12657 behavior). When true, cleanup is offloaded to the
utility thread as introduced by WT-12657.

When use_thread=false and the btree is row-store or VLCS, the inline
checkpoint cleanup branch deliberately omits WT_READ_CACHE so the walk
can pull internal pages from disk and fast-delete obsolete leaves.
During recovery, closing-checkpoint, or RTS we set internal_cleanup to
false but must still limit reads to cache only; otherwise a fast-
truncated leaf can be instantiated from disk and dirty the tree,
tripping the dirty-on-shutdown assertion in __wt_tree_modify_set.

Also regenerate dist-generated files with the older Python version used
on this branch.

kommiharibabu committed:
WT-12657 Add test_cc06 exercising checkpoint_cleanup.use_thread

Scenarios cover both modes:
  - use_thread=false: a plain checkpoint drives cleanup inline, so
    cc_handle_processed > 0 immediately after the call and cc_success
    stays at 0 because the dedicated thread is not started.
  - use_thread=true: a plain checkpoint resets cc_handle_processed to 0
    and nothing bumps it. Triggering the thread via
    debug=(checkpoint_cleanup=true) eventually bumps cc_success and
    cc_handle_processed.

kommiharibabu committed:
WT-12657 Fix test_cc01-05 and test_compact12 hang with use_thread=false default

wait_for_cc_to_run() waits for cc_success to increment, but cc_success
is only bumped by the dedicated cleanup thread. With use_thread=false
(the new default), no thread is started so the wait loops forever.
Add checkpoint_cleanup=(use_thread=true,wait=60) to test_cc01 through
test_cc05 and test_compact12 so the thread-based path these tests were
written for is explicitly enabled.

etienneptl commented on dist/api_data.py:
```diff
@@ -546,10 +546,21 @@ def __ge__(self, other):
+            type='boolean'),
+        Config('wait', '300', r'''
+            seconds to wait between each checkpoint cleanup''',
+            min='60', max='100000'),
```
@kommiharibabu, should we make this 1? We have done this already in develop.

kommiharibabu replied:
Yes, fixed.

kommiharibabu committed:
set the min time to 1 second

kommiharibabu committed:
WT-12657 Fix page_only flag in obsolete cleanup when called inline during checkpoint

When __sync_obsolete_deleted_cleanup and __sync_obsolete_disk_cleanup are
called from the dedicated cleanup thread, page_only=false is correct: both
the parent page and the tree must be marked dirty so the next checkpoint
picks up the change.

When called inline during the checkpoint tree walk (WT_SESSION_BTREE_SYNC
is set), only the parent page needs to be marked dirty — the current
checkpoint will write it immediately. Passing page_only=false in this
context incorrectly calls __wt_tree_modify_set from within the checkpoint
session, which was never intended and sets btree->modified and
conn->modified unnecessarily.

Use WT_SESSION_BTREE_SYNC(session) as the page_only argument so the
correct behavior is selected automatically at both call sites.

kommiharibabu committed:
fix s_all

etienneptl commented on dist/api_data.py:
```diff
@@ -546,10 +546,21 @@ def __ge__(self, other):
+            control how aggressively obsolete content is removed by reading the internal pages.
+            Default to none, which means no additional work is done to find obsolete content.
+            ''', choices=['none', 'reclaim_space']),
+        Config('use_thread', 'false', r'''
+            if true, a dedicated utility thread performs checkpoint cleanup asynchronously.
+            When false (default), cleanup runs inline during the checkpoint tree walk.''',
+            type='boolean'),
```
@kommiharibabu, what if users use that config and when they upgrade, this config no longer exists, will this create some upgrade/downgrade issue?

kommiharibabu replied:
Yes, it can cause the issue when they upgrade, but the TS should communicate the need to remove the new config whenever the respective users who set this new config is set.

etienneptl replied:
How would the config be removed once it's set, wouldn't be written as `use_thread=false`? I think we should go though the exercise to make sure we can "remove it" from the config before upgrading.

kommiharibabu replied:
This new configuration is not saved anywhere in the database files; the customer needs to pass it during the start of the server, so whenever they upgrade, they can just remove it from their config that is passed to the server.

etienneptl committed:
s_all new line

etienneptl reviewed (approved):
LGTM
